include/alpaka/queue/QueueCudaRtSync.hpp

```diff
@@ -247,19 +269,37 @@ namespace alpaka
                         pCallbackSynchronizationData.get(),
                         0u));
 
-                    // If the callback has not yet been called, we wait for it.
-                    std::unique_lock<std::mutex> lock(pCallbackSynchronizationData->m_mutex);
-                    if(!pCallbackSynchronizationData->notified)
-                    {
-                        pCallbackSynchronizationData->m_event.wait(
-                            lock,
-                            [pCallbackSynchronizationData](){
-                                return pCallbackSynchronizationData->notified;
+                    // We start a new std::thread which stores the task to be executed.
+                    // This circumvents the limitation that it is not possible to call CUDA methods within the CUDA callback thread.
+                    // The CUDA thread signals the std::thread when it is ready to execute the task.
+                    // The CUDA thread is waiting for the std::thread to signal that it is finished executing the task
+                    // before it executes the next task in the queue (CUDA stream).
+                    std::thread t(
+                        [pCallbackSynchronizationData, task](){
+
+                            // If the callback has not yet been called, we wait for it.
+                            {
+                                std::unique_lock<std::mutex> lock(pCallbackSynchronizationData->m_mutex);
+                                if(pCallbackSynchronizationData->state != CallbackState::notified)
+                                {
+                                    pCallbackSynchronizationData->m_event.wait(
+                                        lock,
+                                        [pCallbackSynchronizationData](){
+                                            return pCallbackSynchronizationData->state == CallbackState::notified;
+                                        }
+                                    );
+                                }
+
+                                task();
+
+                                // Notify the waiting CUDA thread.
+                                pCallbackSynchronizationData->state = CallbackState::finished;
                             }
-                        );
-                    }
+                            pCallbackSynchronizationData->m_event.notify_one();
+                        }
+                    );
 
-                    task();
+                    t.join();
```

Comment:
The code here is not new but only copied from the `QueueCudaRtAsync`. This had been forgotten when a bug had been fixed there.